concordium-contracts-common-derive/src/derive.rs

```diff
@@ -279,24 +326,145 @@ impl ContainerAttributes {
+    /// can represent the number of variants.
+    fn tag_repr(&self, variants_len: usize) -> syn::Result<syn::Ident> {
+        if let Some(repr_attribute) = &self.repr {
+            let max = repr_attribute.max_variant_length();
+            let ident = repr_attribute.ident();
+            ensure!(
+                max >= variants_len,
+                Span::call_site(),
+                "Too many variants. Maximum {max} are supported with 'repr({ident})'",
+            );
+            Ok(ident)
+        } else if variants_len <= usize::from(u8::MAX) + 1 {
+            Ok(format_ident!("u8"))
+        } else if variants_len <= usize::from(u16::MAX) + 1 {
+            Ok(format_ident!("u16"))
+        } else {
+            bail!(
+                Span::call_site(),
+                "[derive(Deserial)]: Too many variants. Maximum 65536 are supported.",
+            );
+        }
+    }
+
+    /// Ensure that when tag attribute is used, a repr attribute is set and this
+    /// repr is large enough to represent the tag.
+    fn check_tag_with_repr(&self, tag_attribute: &Option<TagAttribute>) -> syn::Result<()> {
+        if let Some(tag) = tag_attribute {
+            let Some(repr) = &self.repr else {
+                bail!(
+                    Span::call_site(),
+                    "'repr(..)' attribute must be set on the type when using 'tag' attribute",
+                );
+            };
+            let value: usize = tag.value.base10_parse()?;
+            ensure!(
+                value <= repr.max_tag_value(),
+                tag.span,
+                "'tag' attribute value of {} cannot be represented by the type {1} set in \
+                 'repr({1})'",
+                tag.value,
+                repr.ident()
+            );
+        }
+        Ok(())
+    }
+}
+
+/// Valid values for the 'repr(..)' attribute.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum ReprAttributeValue {
+    U8,
+    U16,
+}
+
+/// 'repr(..)' attribute, used to specify the size length used for enum tags.
+#[derive(Debug)]
+struct ReprAttribute {
+    /// The size length to use.
+    value: ReprAttributeValue,
+    /// The span of the attribute.
+    span:  proc_macro2::Span,
+}
+
+impl ReprAttribute {
+    /// Construct a ['syn::Ident'] from the value using the attribute span.
+    fn ident(&self) -> syn::Ident {
+        use ReprAttributeValue::*;
+        match self.value {
+            U8 => format_ident!("u8", span = self.span),
+            U16 => format_ident!("u16", span = self.span),
+        }
+    }
+
+    /// Maximum tag value with the value of the 'repr' attribute.
+    fn max_tag_value(&self) -> usize {
+        use ReprAttributeValue::*;
+        match self.value {
+            U8 => u8::MAX.into(),
+            U16 => u16::MAX.into(),
+        }
+    }
+
+    /// Maximum number of variants distiguishable with the value of the 'repr'
+    /// attribute
+    fn max_variant_length(&self) -> usize { self.max_tag_value() + 1 }
```

Comment:
```suggestion
    fn max_num_variants(&self) -> usize { self.max_tag_value() + 1 }
```

That fits exactly the description above.